open_flamingo/train/train_utils.py

```diff
@@ -298,41 +221,132 @@ def update(self, val, n=1):
+            resume_from_checkpoint = args.run_name
+            print(f"Found checkpoint {resume_from_checkpoint} for run {args.run_name}.")
+        else:
+            print(f"Found no checkpoints for run {args.run_name}.")
+    else:
+        checkpoint_list = glob.glob(f"{args.run_name}/checkpoint_*.pt")
+        if len(checkpoint_list) == 0:
+            print(f"Found no checkpoints for run {args.run_name}.")
+        else:
+            resume_from_checkpoint = sorted(
+                checkpoint_list, key=lambda x: int(x.split("_")[-1].split(".")[0])
+            )[-1]
+            print(f"Found checkpoint {resume_from_checkpoint} for run {args.run_name}.")
+    return resume_from_checkpoint
+
+
+def load_checkpoint(args, model):
+    """
+    Loads a (non-Deepspeed) checkpoint into the model and returns the checkpoint + epoch to resume from.
+    Does not load the optimizer or learning rate checkpoints, but these are included in the returned checkpoint dict.
+    """
+    if args.rank == 0:
+        print(f"Loading checkpoint from {args.resume_from_checkpoint}")
+    checkpoint = torch.load(args.resume_from_checkpoint, map_location="cpu")
+    msd = checkpoint.pop("model_state_dict")
+    msd = {k.replace("module.", ""): v for k, v in msd.items()}
+    resume_from_epoch = checkpoint["epoch"] + 1
+    if args.fsdp:
+        FSDP.set_state_dict_type(
+            model,
+            **args.fsdp_checkpoint_config,
+        )
+    model.load_state_dict(msd, False)
+    return resume_from_epoch, checkpoint
+
+
+def load_deepspeed_checkpoint(args, ddp_model):
+    """Loads a deepspeed checkpoint and returns the epoch to resume from."""
+    if args.rank == 0:
+        print(f"Loading checkpoint from {args.resume_from_checkpoint}")
+    # We will not pass in a 'tag' and instead rely on 'latest' file in the checkpoint directory
+    ddp_model.load_checkpoint(
+        load_dir=args.resume_from_checkpoint,  # Note: this is the dir, not the file
+        load_module_strict=False,
+    )
+    # read latest file to get epoch
+    latest_file = os.path.join(args.resume_from_checkpoint, "latest")
+    with open(latest_file, "r") as f:
+        checkpoint_epoch = int(f.read().split("_")[-1])
+    return checkpoint_epoch + 1
+
+
 def filter_state_dict_to_trainable(model, state_dict):
     """
     Remove non-trainable parameters from model state dict.
     Exception: Embeddings will not be removed, even if frozen.
     This is because we need the new <image> <|endofchunk|> tokens to
     be consistent across initializations.
     """
-    for (
-        name,
-        p,
-    ) in model.named_parameters():  # won't work for fsdp + use_orig_params=False
+    # first, remove frozen params
+    for name, p in model.named_parameters():
         if "fsdp" in name:
             continue
-        if "embed" in name or isinstance(p, torch.nn.Embedding):
-            continue
-        if not p.requires_grad:
+        if not p.requires_grad or to_delete(name):
```

Comment:
Looks like to_delete isn't defined anywhere?

Reply:
Ah! You're right, let me push a fix.